:bug:issue-1737: Add unit tests for openstackmachine_webhook

MykolaRodin opened the pull request:
**What this PR does / why we need it**:

It extends tests for `openstackmachine_webhook`

Fixes: #1737

mdbooth commented on test/e2e/suites/apivalidations/openstackmachine_test.go:
```diff
@@ -50,7 +50,7 @@ var _ = Describe("OpenStackMachine API validations", func() {
 	})
 
-	It("should only allow the providerID to be set once", func() {
+	It("should only allow the providerID to be set once and identityRef to be set several times", func() {
```
Please can you add a separate test for identityRef instead of adding it to this one? Unlike the e2e tests these tests are pretty fast to initialise, so it's ok to run lots of them. It's simpler to understand and maintain if they each do one thing only.

mdbooth commented on test/e2e/suites/apivalidations/openstackmachine_test.go:
```diff
@@ -63,6 +63,14 @@ var _ = Describe("OpenStackMachine API validations", func() {
 		Expect(k8sClient.Update(ctx, machine)).NotTo(Succeed(), "Updating providerID should fail")
+
+		By("Setting the identityRef")
+		machine.Spec.IdentityRef = ptr.To(infrav1.OpenStackIdentityReference{Name: "foo", CloudName: "GCP"})
```
Incidentally, it's not important to this test, except for readability, but `CloudName` here refers to an entry for an OpenStack cloud in `clouds.yaml`. So it's going to be something like `production`, `staging`, etc. In practise people don't tend to push a clouds.yaml with more than 1 cloud in it so this is academic, but we still support it.

mdbooth commented on test/e2e/suites/apivalidations/openstackmachine_test.go:
```diff
@@ -63,6 +63,14 @@ var _ = Describe("OpenStackMachine API validations", func() {
+
+		// By("Modifying the identityRef")
+		// machine.Spec.IdentityRef = ptr.To(infrav1.OpenStackIdentityReference{Name: "bar", CloudName: "AWS"})
+		// Expect(k8sClient.Update(ctx, machine)).To(Succeed(), "Updating identityRef should succeed")
```
Is it this one that fails?

MykolaRodin replied:
I have not uncommented it yet because it fails before.

If I change `providerID` and `identityRef` in different tests, it works fine. If I change both values withing the same test, if fails. I updated `cluster-api-provider-openstack/pkg/webhooks/openstackmachine_webhook.go` - `ValidateUpdate()` a bit to print old and new objects on fail. Here the results of the fail are:
```
Message: "admission webhook \"validation.openstackmachine.infrastructure.cluster.x-k8s.io\" denied the request: OpenStackMachine.infrastructure.cluster.x-k8s.io \"machine-vjrtn\" is invalid: spec: Forbidden: cannot be modified

old:map[flavor: image:map[filter:map[name:test-image]] providerID:foo]

new:map[flavor: image:map[filter:map[name:test-image]] providerID:bar]"
```
Please clarify if it is expected.

MykolaRodin replied:
OK, it looks like I got the reason for the failure when updating both `providerID` and `identityRef`. The test failed because updating `providerID` for the second time failed. Once I removed it, the test with updating `providerID` once and updating `identityRef` twice succeeded. Please clarify whether this test with updating both `providerID` and `identityRef` should be kept or it should be deleted

mdbooth commented:
/ok-to-test

MykolaRodin commented:
/retest

MykolaRodin commented:
HI @mdbooth,

I have removed the test with updating providerID once and updating identityRef twice. I have removed my temporary changes in `cluster-api-provider-openstack/pkg/webhooks/openstackmachine_webhook.go` - `ValidateUpdate()` and left just a comment there that should help to understand why `instanceID` is mentioned there. The unit tests that I added beside the e2e test work fine for me. Please clarify what else I am expected to implement in this MR?

mdbooth commented on test/e2e/suites/apivalidations/openstackmachine_test.go:
```diff
@@ -65,6 +65,21 @@ var _ = Describe("OpenStackMachine API validations", func() {
 		Expect(k8sClient.Update(ctx, machine)).NotTo(Succeed(), "Updating providerID should fail")
 	})
 
+	It("should only allow the identityRef to be set several times", func() {
```
Nit:
```suggestion
	It("should allow the identityRef to be set several times", func() {
```

MykolaRodin replied:
Done

mdbooth commented on pkg/webhooks/openstackmachine_webhook.go:
```diff
@@ -96,6 +96,7 @@ func (*openStackMachineWebhook) ValidateUpdate(_ context.Context, oldObjRaw, new
 		delete(newOpenStackMachineSpec, "providerID")
 	}
 
+	// instanceID is present in v1alpha6 but is deprecated in v1beta1
```
This is the problem of writing these webhooks with `Unstructured` instead of proper types. The code below is a no-op because `instanceID` can never be set. If we were using proper types it would have been a compile error and we'd have noticed.

This can be deleted. Probably best in a separate PR, though, or at least a separate commit.

MykolaRodin replied:
`// instanceID is present in v1alpha6 but is deprecated in v1beta1` was removed

mdbooth commented on pkg/webhooks/openstackmachine_webhook_test.go:
```diff
@@ -0,0 +1,197 @@
+			wantErr: false,
+		},
+		{
+			name: "OpenStackMachine.Spec.RootVolume and OpenStackMachine.Spec.AdditionalBlockDevices with non-root user on create",
```
```suggestion
			name: "OpenStackMachine.Spec.RootVolume and OpenStackMachine.Spec.AdditionalBlockDevices with non-root device name on create",
```

MykolaRodin replied:
Done

mdbooth commented on pkg/webhooks/openstackmachine_webhook_test.go:
```diff
@@ -0,0 +1,197 @@
+			wantErr: false,
+		},
+		{
+			name: "OpenStackMachine.Spec.RootVolume and OpenStackMachine.Spec.AdditionalBlockDevices with root user on create",
```
```suggestion
			name: "OpenStackMachine.Spec.RootVolume and OpenStackMachine.Spec.AdditionalBlockDevices with root device name on create",
```

MykolaRodin replied:
Done

mdbooth commented on pkg/webhooks/openstackmachine_webhook_test.go:
These tests are good. Ideally they'd be implemented in the apivalidations suite instead, though, because in the medium term I actually want to delete this webhook entirely and implement these validations in the apiserver directly. When we do that, these tests will no longer pass.

However, if we implement these tests in apivalidations instead we can completely change the implementation of the validations while ensuring that the exact same test suite continues to pass.

If you are able, my preference would be to move them to apivalidations. If not, though, I'll take these anyway. Apart from anything else they document the test cases that need to be implemented in apivalidations when we change the validations.

MykolaRodin replied:
Thank you for the review. I have updated my PR in accordance with your code review remarks. On the other hand, I am sorry I cannot get what you mean under moving the tests to `apivalidations`. The e2e test I added has already been placed in `cluster-api-provider-openstack/test/e2e/suites/apivalidations` folder. The unit tests I added are placed in the file in the same folder as `cluster-api-provider-openstack/pkg/webhooks/openstackmachine_webhook.go` file. I cannot find other `apivalidations` folder and cannot see what I cad additionally change in the paths.

If it is possible, I would like to merge this PR as is. If you provide me with the additional clarification related to moving the tests to `apivalidations`, I am ready to do it in a separate PR.

MykolaRodin replied:
Could you please clarify if I should have added several e2e tests instead of the unit tests?

mdbooth replied:
My *preference* was that they be e2e tests instead. However these tests improve the current state of our testing and I'll gladly merge unit tests.

MykolaRodin replied:
Hi @mdbooth,

It looks like still PR should still be modified to be taken into account and be reviewed. I removed the unit tests that I added and substituted them with e2e tests. I hope it is a bit closer to your preferences. Please review it and let me know if I am still missing something.

mdbooth commented:
/approve

These are great, thanks! If you wanted to convert the unit tests to be e2e tests instead that would also be great, but I'm happy for these to merge as-is.

mdbooth commented:
FYI, I added:
```
Fixes: #1737
```
to the issue description so the issue will be automatically closed when this PR merges.

mdbooth reviewed (commented):
This is great, thanks. I have already approved. Lets wait for an lgtm from another reviewer.

lentzi90 reviewed (commented):
/lgtm

MykolaRodin commented:
Hi @mdbooth ,

The tests failed with this error:
```
   [FAILED] Timed out after 1800.001s.
  Timed out waiting for 3 control plane machines to exist
  Expected
      <int>: 2
  to equal
      <int>: 3
  In [BeforeEach] at: /root/go/pkg/mod/sigs.k8s.io/cluster-api/test@v1.7.0/framework/controlplane_helpers.go:116 @ 05/10/24 11:31:12.929
  Full Stack Trace
    sigs.k8s.io/cluster-api/test/framework.WaitForKubeadmControlPlaneMachinesToExist({0x2ac29b8, 0x4033060}, {{0x7f2068780ca0?, 0xc000b122d0?}, 0xc00122c1a0?, 0xc000613208?}, {0xc000f695a0, 0x2, 0x2})
    	/root/go/pkg/mod/sigs.k8s.io/cluster-api/test@v1.7.0/framework/controlplane_helpers.go:116 +0x33b
    sigs.k8s.io/cluster-api/test/framework.WaitForControlPlaneAndMachinesReady({0x2ac29b8, 0x4033060}, {{0x7f2068758ad0?, 0xc000b122d0?}, 0xc00122c1a0?, 0xc000613208?}, {0xc000f695a0, 0x2, 0x2})
    	/root/go/pkg/mod/sigs.k8s.io/cluster-api/test@v1.7.0/framework/controlplane_helpers.go:287 +0x692
    sigs.k8s.io/cluster-api/test/framework/clusterctl.ApplyCustomClusterTemplateAndWait.setDefaults.func3({_, _}, {{0x2ad4d18, 0xc000801800}, {0xc00100a001, 0x40529, 0x4052a}, {0xc00093c7e0, 0x12}, {0xc000c85590, ...}, ...}, ...) 
```
I am sorry, I am not quite sure what it actually means. `make test` runs successfully on my local machine.

lentzi90 commented:
It is an e2e test that failed. They are a bit flaky sometimes unfortunately.
As you only touched unit test code there cannot be any relation to these changes. I'll retrigger
/test pull-cluster-api-provider-openstack-e2e-test

EmilienM reviewed (approved):
/lgtm

EmilienM commented:
/hold waiting for the CI job

mdbooth commented:
/hold cancel
